packages/flutter/lib/src/rendering/table.dart

```diff
@@ -385,76 +385,48 @@ class RenderTable extends RenderBox {
        _configuration = configuration {
-    _children = <RenderBox?>[]..length = _columns * _rows;
     this.rowDecorations = rowDecorations; // must use setter to initialize box painters array
-    children?.forEach(addRow);
+    if (children != null) {
+      final List<RenderBox> allChildren = <RenderBox>[];
+      for (final List<RenderBox> rows in children)
+        rows.forEach(allChildren.add);
+
+      allChildren.reversed.forEach(insertChild);
+    }
   }
 
   // Children are stored in row-major order.
   // _children.length must be rows * columns
-  List<RenderBox?> _children = const <RenderBox?>[];
+  final List<RenderBox?> _children = <RenderBox?>[];
 
   /// The number of vertical alignment lines in this table.
   ///
-  /// Changing the number of columns will remove any children that no longer fit
-  /// in the table.
-  ///
-  /// Changing the number of columns is an expensive operation because the table
-  /// needs to rearrange its internal representation.
+  /// This can be updated by [setDimensions] after updating children is finish.
   int get columns => _columns;
   int _columns;
-  set columns(int value) {
-    assert(value != null);
-    assert(value >= 0);
-    if (value == columns)
-      return;
-    final int oldColumns = columns;
-    final List<RenderBox?> oldChildren = _children;
-    _columns = value;
-    _children = List<RenderBox?>.filled(columns * rows, null, growable: false);
-    final int columnsToCopy = math.min(columns, oldColumns);
-    for (int y = 0; y < rows; y += 1) {
-      for (int x = 0; x < columnsToCopy; x += 1)
-        _children[x + y * columns] = oldChildren[x + y * oldColumns];
-    }
-    if (oldColumns > columns) {
-      for (int y = 0; y < rows; y += 1) {
-        for (int x = columns; x < oldColumns; x += 1) {
-          final int xy = x + y * oldColumns;
-          if (oldChildren[xy] != null)
-            dropChild(oldChildren[xy]!);
-        }
-      }
-    }
-    markNeedsLayout();
-  }
 
   /// The number of horizontal alignment lines in this table.
   ///
-  /// Changing the number of rows will remove any children that no longer fit
-  /// in the table.
+  /// This can be update by [setDimensions] after updating children is finish.
   int get rows => _rows;
   int _rows;
-  set rows(int value) {
-    assert(value != null);
-    assert(value >= 0);
-    if (value == rows)
+
+  /// Update the number of row and column of this table.
+  ///
+  /// This method should be called after updating children is finish.
+  void setDimensions(int rows, int columns) {
+    assert(_children.length == rows * columns);
+    if (rows == _rows && columns == _columns)
       return;
-    if (_rows > value) {
-      for (int xy = columns * value; xy < _children.length; xy += 1) {
-        if (_children[xy] != null)
-          dropChild(_children[xy]!);
-      }
-    }
-    _rows = value;
-    _children.length = columns * rows;
+    _columns = columns;
+    _rows = rows;
     markNeedsLayout();
```

Comment:
this doesn't seem to handle _children being too long compared to the number of rows and columns any more